tests/std/tests/P2445R1_forward_like/test.compile.pass.cpp

```diff
@@ -27,52 +27,94 @@ constexpr owner_type owner_c{};
-
-static_assert(is_same_v<decltype(forward_like<decltype(owner_lref)>(owner_lref.m_cobj)), const int&>);
-static_assert(is_same_v<decltype(forward_like<decltype(owner_lref)>(owner_lref.m_clref)), const int&>);
-static_assert(is_same_v<decltype(forward_like<decltype(owner_lref)>(owner_lref.m_crref)), const int&>);
-
-static_assert(is_same_v<decltype(forward_like<decltype(owner_rref)>(owner_rref.m_mutobj)), int&&>);
-static_assert(is_same_v<decltype(forward_like<decltype(owner_rref)>(owner_rref.m_mutlref)), int&&>);
-static_assert(is_same_v<decltype(forward_like<decltype(owner_rref)>(owner_rref.m_mutrref)), int&&>);
-
-static_assert(is_same_v<decltype(forward_like<decltype(owner_rref)>(owner_rref.m_cobj)), const int&&>);
-static_assert(is_same_v<decltype(forward_like<decltype(owner_rref)>(owner_rref.m_clref)), const int&&>);
-static_assert(is_same_v<decltype(forward_like<decltype(owner_rref)>(owner_rref.m_crref)), const int&&>);
-
-static_assert(is_same_v<decltype(forward_like<decltype(owner_c)>(owner_c.m_mutobj)), const int&&>);
-static_assert(is_same_v<decltype(forward_like<decltype(owner_c)>(owner_c.m_mutlref)), const int&&>);
-static_assert(is_same_v<decltype(forward_like<decltype(owner_c)>(owner_c.m_mutrref)), const int&&>);
-
-static_assert(is_same_v<decltype(forward_like<decltype(owner_c)>(owner_c.m_cobj)), const int&&>);
-static_assert(is_same_v<decltype(forward_like<decltype(owner_c)>(owner_c.m_clref)), const int&&>);
-static_assert(is_same_v<decltype(forward_like<decltype(owner_c)>(owner_c.m_crref)), const int&&>);
-
-static_assert(is_same_v<decltype(forward_like<decltype(owner_clref)>(owner_clref.m_mutobj)), const int&>);
-static_assert(is_same_v<decltype(forward_like<decltype(owner_clref)>(owner_clref.m_mutlref)), const int&>);
-static_assert(is_same_v<decltype(forward_like<decltype(owner_clref)>(owner_clref.m_mutrref)), const int&>);
-
-static_assert(is_same_v<decltype(forward_like<decltype(owner_clref)>(owner_clref.m_cobj)), const int&>);
-static_assert(is_same_v<decltype(forward_like<decltype(owner_clref)>(owner_clref.m_clref)), const int&>);
-static_assert(is_same_v<decltype(forward_like<decltype(owner_clref)>(owner_clref.m_crref)), const int&>);
-
-static_assert(is_same_v<decltype(forward_like<decltype(owner_crref)>(owner_crref.m_mutobj)), const int&&>);
-static_assert(is_same_v<decltype(forward_like<decltype(owner_crref)>(owner_crref.m_mutlref)), const int&&>);
-static_assert(is_same_v<decltype(forward_like<decltype(owner_crref)>(owner_crref.m_mutrref)), const int&&>);
-
-static_assert(is_same_v<decltype(forward_like<decltype(owner_crref)>(owner_crref.m_cobj)), const int&&>);
-static_assert(is_same_v<decltype(forward_like<decltype(owner_crref)>(owner_crref.m_clref)), const int&&>);
-static_assert(is_same_v<decltype(forward_like<decltype(owner_crref)>(owner_crref.m_crref)), const int&&>);
+using owner_t              = decltype(owner);

```

Comment:
I have some concerns about test coverage. The expressions `owner`, `owner_lref`, and `owner_rref` are all lvalues of type `owner_type`. Given an lvalue `x` of type `owner_type`, the expressions `x.m_mutobj`, `x.m_mutlref`, and `x.m_mutrref` are all lvalues of type `int`. We pass an lvalue of type `int` to `forward_like<owner_t>`, `forward_like<owner_lref_t>`, and `forward_like<owner_rref_t>` three times each. This seems like overkill when once each would suffice. We similarly pass lvalues of type `const int` three times each to `forward_like<owner_t>`, `forward_like<owner_lref_t>`, and `forward_like<owner_rref_t>`.

For the `const` paths, the numbers are a little different. The expressions `owner_c`, `owner_clref`, and `owner_crref` are all lvalues of type `const owner_type`. Given an lvalue `x` of type `const owner_type`, the expressions `x.m_mutobj`, `x.m_cobj`, `x.m_clref` and `x.m_crref` are all lvalues of type `const int`. The expressions `x.m_mutlref` and `x.m_mutrref` are lvalues of type `int`. So instead of 3 `const` and 3 non-`const` lvalue `int`s per set of calls we have 4 `const` and 2 non-`const` lvalue `int`s per set of calls.

I think the `owner` class and the complicated naming is all obfuscating what's happening here and the fact that we are missing coverage despite having quite a bit of redundant coverage. I think we could greatly simplify all of this to:
```c+
struct T {}; // class type so const prvalues are valid
using CT = const T;
using U  = int;
using CU = const U;

T t{};
const T& ct = t;

static_assert(is_same_v<decltype(forward_like<U>(T{})), T&&>);
static_assert(is_same_v<decltype(forward_like<U>(CT{})), CT&&>);
static_assert(is_same_v<decltype(forward_like<U>(t)), T&&>);
static_assert(is_same_v<decltype(forward_like<U>(ct)), CT&&>);
static_assert(is_same_v<decltype(forward_like<U>(move(t))), T&&>);
static_assert(is_same_v<decltype(forward_like<U>(move(ct))), CT&&>);

static_assert(is_same_v<decltype(forward_like<CU>(T{})), CT&&>);
static_assert(is_same_v<decltype(forward_like<CU>(CT{})), CT&&>);
static_assert(is_same_v<decltype(forward_like<CU>(t)), CT&&>);
static_assert(is_same_v<decltype(forward_like<CU>(ct)), CT&&>);
static_assert(is_same_v<decltype(forward_like<CU>(move(t))), CT&&>);
static_assert(is_same_v<decltype(forward_like<CU>(move(ct))), CT&&>);

static_assert(is_same_v<decltype(forward_like<U&>(T{})), T&>);
static_assert(is_same_v<decltype(forward_like<U&>(CT{})), CT&>);
static_assert(is_same_v<decltype(forward_like<U&>(t)), T&>);
static_assert(is_same_v<decltype(forward_like<U&>(ct)), CT&>);
static_assert(is_same_v<decltype(forward_like<U&>(move(t))), T&>);
static_assert(is_same_v<decltype(forward_like<U&>(move(ct))), CT&>);

static_assert(is_same_v<decltype(forward_like<CU&>(T{})), CT&>);
static_assert(is_same_v<decltype(forward_like<CU&>(CT{})), CT&>);
static_assert(is_same_v<decltype(forward_like<CU&>(t)), CT&>);
static_assert(is_same_v<decltype(forward_like<CU&>(ct)), CT&>);
static_assert(is_same_v<decltype(forward_like<CU&>(move(t))), CT&>);
static_assert(is_same_v<decltype(forward_like<CU&>(move(ct))), CT&>);

static_assert(is_same_v<decltype(forward_like<U&&>(T{})), T&&>);
static_assert(is_same_v<decltype(forward_like<U&&>(CT{})), CT&&>);
static_assert(is_same_v<decltype(forward_like<U&&>(t)), T&&>);
static_assert(is_same_v<decltype(forward_like<U&&>(ct)), CT&&>);
static_assert(is_same_v<decltype(forward_like<U&&>(move(t))), T&&>);
static_assert(is_same_v<decltype(forward_like<U&&>(move(ct))), CT&&>);

static_assert(is_same_v<decltype(forward_like<CU&&>(T{})), CT&&>);
static_assert(is_same_v<decltype(forward_like<CU&&>(CT{})), CT&&>);
static_assert(is_same_v<decltype(forward_like<CU&&>(t)), CT&&>);
static_assert(is_same_v<decltype(forward_like<CU&&>(ct)), CT&&>);
static_assert(is_same_v<decltype(forward_like<CU&&>(move(t))), CT&&>);
static_assert(is_same_v<decltype(forward_like<CU&&>(move(ct))), CT&&>);
```
`|{const, non-const} x {prvalue, lvalue, xvalue}| == 6`, so we have precisely 6 sets of 6 test cases each to get full coverage. Does that make sense?

Reply:
Thank you very much! I used to be considering "owner and member" in the paper, and be afraid that `int` is not suitable as an "owner". But I can go with it now because `int` and `T` are capable to emulate the "far" state.

I'm taking your test coverage, making the comments a bit wordy but more clear to me.